### DISCUSSION TITLE
Special functions for expression evaluations

### DISCUSSION THREAD

**Discussion by ghalimi** [stdlib-js/stdlib#379] [Q&A] [answered]
The [exprtk](https://github.com/ArashPartow/exprtk) C++ Mathematical Expression Toolkit Library makes a clever use of [special functions](https://github.com/ArashPartow/exprtk/blob/master/readme.txt#L1510) to accelerate the evaluation of expressions. Is there an opportunity for stdlib to do the same, or does it have to be done with some pre-compiler?

**Comment by kgryte** [comment, marked as answer]
exprtk could be useful as part of a CAS; however, in terms of reducing the number of AST nodes for purposes of perf improvements, this is not always desirable, as these optimizations may lead to accuracy degradations. Given IEEE 754 floating-point format, operations which are theoretically commutative are not always commutative in practice. So anything which performs simplifications, reorderings, etc, may violate an algorithm's assumptions and result in decreased accuracy. It is not uncommon for transcendental function algorithms to assume (and exploit) certain properties of IEEE 754 to achieve results to a desired precision.

This should also be a more general warning when translating/porting numerical algorithms: the order of operations matters and any change to that order can result in significant deviations in accuracy.
